Goal: Use slapd.d "runtime configuration" for slapd

Fixes  #596710 (kolabd: kolab_bootstrap no longer works with
slapd (version 2.4.23))

Author: Mathieu Parent <math.parent@gmail.com>

Upstream status: https://issues.kolab.org/issue3000

--- a/lib/Skolab/Conf.pm
+++ b/lib/Skolab/Conf.pm
@@ -649,6 +649,17 @@
         if ($args{doreload} || $cmd !~ /(^invoke-rc.d| rc) \S+ re(start|load)/) {
             Skolab::log('T', 'Executing command: '.$cmd, SKOLAB_DEBUG );
             system($cmd);
+        } elsif ($args{doreload} || $cmd !~ /(^invoke-rc.d| rc) (slapd|openldap) re(start|load)/) {
+            &Skolab::log('S', 'Stopping OpenLDAP...');
+            system("invoke-rc.d slapd stop");
+            &Skolab::log('S', 'Deleting slapd.d/ config...');
+            system("rm -rf $Skolab::config{'ldapserver_confdir'}/slapd.d");
+            &Skolab::log('S', 'Converting slapd config... (most errors here can be ignored)');
+            system("mkdir $Skolab::config{'ldapserver_confdir'}/slapd.d");
+            system("slaptest -f $Skolab::config{'ldapserver_confdir'}/slapd.conf -F $Skolab::config{'ldapserver_confdir'}/slapd.d");
+            system("chown -R openldap:openldap $Skolab::config{'ldapserver_confdir'}/slapd.d");
+            &Skolab::log('S', 'Starting OpenLDAP...');
+            system("invoke-rc.d slapd start");
         } else {
             Skolab::log('T', 'Reload not allowed, not executing command: '.$cmd, SKOLAB_DEBUG );
         }
--- a/sbin/skolab_bootstrap.in
+++ b/sbin/skolab_bootstrap.in
@@ -521,6 +521,14 @@
         );
       $Skolab::config{"bootstrap_config"} = 'false';
 
+      print "Deleting old slapd config...\n";
+      system("rm -rf $Skolab::config{'ldapserver_confdir'}/slapd.d");
+      print "Converting slapd config...\n";
+      system("mkdir $Skolab::config{'ldapserver_confdir'}/slapd.d");
+      system("slaptest -f $Skolab::config{'ldapserver_confdir'}/slapd.conf -F $Skolab::config{'ldapserver_confdir'}/slapd.d");
+      system("chown -R openldap $Skolab::config{'ldapserver_confdir'}/slapd.d");
+      system("chgrp -R openldap $Skolab::config{'ldapserver_confdir'}/slapd.d");
+
       #ldap server should access to certificate key
       system("adduser --quiet $Skolab::config{'ldapserver_rusr'} $Skolab::config{'pki_grp'}");
 
@@ -530,7 +538,7 @@
       # ensure that the database has correct permissions
       @files = <$Skolab::config{'ldapserver_dir'}/*>;
       system("chown $Skolab::config{'ldapserver_rusr'}:$Skolab::config{'ldapserver_grp'} $Skolab::config{'ldapserver_dir'}/*") if (@files);
-      (system("/usr/sbin/slapd -h ldap://127.0.0.1:389/ -f $Skolab::config{'ldapserver_confdir'}/slapd.conf -u $Skolab::config{'ldapserver_rusr'} -g $Skolab::config{'ldapserver_grp'}") == 0 ) || die( "Could not start temporary slapd: $!" );
+      (system("/usr/sbin/slapd -h ldap://127.0.0.1:389/ -f $Skolab::config{'ldapserver_confdir'}/slapd.d -u $Skolab::config{'ldapserver_rusr'} -g $Skolab::config{'ldapserver_grp'}") == 0 ) || die( "Could not start temporary slapd: $!" );
       print ("Waiting for OpenLDAP to start\n");
       sleep 10;
 
